httpio_async/__init__.py

```diff
@@ -0,0 +1,297 @@
+file like objects of python) except that many methods are replaced with asynchronous coroutines."""
+
+import aiohttp
+from httpio import HTTPIOError
+
+
+__all__ = ["AsyncHTTPIOFile", "HTTPIOError", "open"]
+
+
+async def open(url, block_size=-1, **kwargs):
+    """
+    Open a URL as an asynchronous file-like object
+
+    The returned object is open, and needs to be closed when done.
+    This will be handled automatically if you use it as an asynchronous context manager.
+
+    :param url: The URL of the file to open
+    :param block_size: The cache block size, or `-1` to disable caching.
+    :param kwargs: Additional arguments to pass to `requests.Request()`
+    :return: An `httpio.HTTPIOFile` object supporting most of the usual
+        file-like object methods.
+    """
+    f = AsyncHTTPIOFile(url, block_size, **kwargs)
+    await f.open()
+    return f
+
+
+class AsyncHTTPIOFile(object):
+    """An asynchronous equivalent to httpio.HTTPIOFile.
+    Sadly this class cannot descend from that one for technical reasons.
+    """
+    def __init__(self, url, block_size=-1, **kwargs):
+        """
+        :param url: The URL of the file to open
+        :param block_size: The cache block size, or `-1` to disable caching.
+        :param kwargs: Additional arguments to pass to `session.get`
+        """
+        super(AsyncHTTPIOFile, self).__init__()
+        self.url = url
+        self.block_size = block_size
+
+        self._kwargs = kwargs
+        self._cursor = 0
+        self._cache = {}
+        self._session = None
+        self._aiter = None
+
+        self.length = None
+        self.closed = False
+
+    def __repr__(self):
+        status = "closed" if self.closed else "open"
+        return "<%s %s %r at %s>" % (status, type(self).__name__, self.url, hex(id(self)))
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if self._aiter is None:
+            self._aiter = self.readlines()
+
+        return await self._aiter.__anext__()
+
+    async def open(self):
+        """This method has no direct equivalent in the HTTPIOFile interface, but does things which
+        are the equivalent of what's done in the constructor of that class. Since constructors cannot
+        be coroutines this class needs this as a seperate coroutine"""
+
+        if self._session is None:
+            self._session = await aiohttp.ClientSession().__aenter__()
+            async with self._session.head(self.url, **self._kwargs) as response:
+                response.raise_for_status()
+                self.length = int(response.headers.get('content-length', None))
+                self.closed = False
+
+    async def __aenter__(self):
+        await self.open()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+
+    async def close(self):
+        if not self.closed:
+            await self._session.__aexit__(None, None, None)
```

Comment:
Add a check here that _session is not None? This is required when open has not been called.

Reply:
If open has not been called then closed will be False.

Reply:
I might be missing something, but it's initialised to `False` at line 52.

Reply:
You're right. I changed it and didn't update this to match. Will fix.